lofo/lofo_importance.py

```diff
@@ -9,72 +9,66 @@
 
 class LOFOImportance:
 
-    def __init__(self, df, features, target,
-                 scoring, model=None, cv=4, n_jobs=None):
+    def __init__(self, dataset, scoring, model=None, fit_params=None, cv=4, n_jobs=None):
 
-        df = df.copy()
-        self.fit_params = {}
+        self.fit_params = fit_params if fit_params else dict()
         if model is None:
-            model, df, categoricals = infer_model(df, features, target, n_jobs)
+            model, dataset.df, categoricals, dataset.y = infer_model(dataset.df, dataset.features, dataset.y, n_jobs)
             self.fit_params["categorical_feature"] = categoricals
             n_jobs = 1
 
         self.model = model
-        self.df = df
-        self.features = features
-        self.target = target
+        self.dataset = dataset
         self.scoring = scoring
         self.cv = cv
         self.n_jobs = n_jobs
         if self.n_jobs is not None and self.n_jobs > 1:
-            warning_str = "Warning: If your model is multithreaded, please initialise the number \
-                of jobs of LOFO to be equal to 1, otherwise you may experience issues."
+            warning_str = ("Warning: If your model is multithreaded, please initialise the number"
+                           "of jobs of LOFO to be equal to 1, otherwise you may experience performance issues.")
             warnings.warn(warning_str)
 
-    def _get_cv_score(self, X, y):
-        fit_params = self.fit_params.copy()
-        if "categorical_feature" in self.fit_params:
-            fit_params["categorical_feature"] = [cat for cat in fit_params["categorical_feature"] if cat in X.columns]
+    def _get_cv_score(self, feature_to_remove):
+        X, fit_params = self.dataset.getX(feature_to_remove=feature_to_remove, fit_params=self.fit_params)
+        y = self.dataset.y
 
         with warnings.catch_warnings():
             warnings.simplefilter("ignore")
             cv_results = cross_validate(self.model, X, y, cv=self.cv, scoring=self.scoring, fit_params=fit_params)
         return cv_results['test_score']
 
-    def _get_cv_score_parallel(self, feature, feature_list, result_queue):
-        test_score = self._get_cv_score(self.df[feature_list], self.df[self.target])
+    def _get_cv_score_parallel(self, feature, result_queue):
+        test_score = self._get_cv_score(feature_to_remove=feature)
         result_queue.put((feature, test_score))
         return test_score
 
     def get_importance(self):
-        base_cv_score = self._get_cv_score(self.df[self.features], self.df[self.target])
+        base_cv_score = self._get_cv_score(feature_to_remove=None)
+        feature_list = self.dataset.features + list(self.dataset.feature_groups.keys())
```

Comment:
Perhaps better to wrap with list(set(...)) in case user replicates features in feature_groups (although maybe we want something to break in case this happens?)

Reply:
feature_groups is a dictionary. Therefore, keys will be unique.

Reply:
Yes I meant when you add them to self.dataset.features

Reply:
Nice catch! I will update it so that dataset will throw exception when it is created